compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -99,6 +188,208 @@ struct DistributeElementwise final : OpDistributionPattern<OpTy> {
   }
 };
 
+/// Given a projected permutation, get a reduced permutation, i.e. without
+/// the projected dimensions.
+static SmallVector<int64_t> getReducedPermutation(AffineMap permutationMap) {
+  assert(permutationMap.isProjectedPermutation() &&
+         "permutation map should be a projected permutation.");
+  // TODO: The permutation map may also have broadcasting. Currently, we do not
+  // handle it. This can be fixed by adding a "BROADCAST" dimension in the
+  // layout.
+
+  SmallVector<int64_t> permutation;
+  permutation.reserve(permutationMap.getNumResults());
+
+  unsigned leadingUnitDims =
+      permutationMap.getNumDims() - permutationMap.getNumResults();
+  for (AffineExpr dim : permutationMap.getResults()) {
+    // Get this dim's position in the permutation map.
+    auto dimExpr = dyn_cast<AffineDimExpr>(dim);
+    if (!dimExpr) {
+      llvm::report_fatal_error("permutation map is not a projected "
+                               "permutation.");
+    }
+
+    unsigned pos = dimExpr.getPosition();
+    assert(pos >= leadingUnitDims && "invalid permutation map");
+    pos -= leadingUnitDims;
+    permutation.push_back(pos);
+  }
+  return permutation;
+}
+
+template <typename OpTy>
+struct DistributeXferLayoutAttr : OpDistributionPattern<OpTy> {
+  static_assert(std::is_same<OpTy, vector::TransferReadOp>::value ||
+                    std::is_same<OpTy, vector::TransferWriteOp>::value,
+                "expected vector::TransferReadOp or vector::TransferWriteOp");
+
+  DistributeXferLayoutAttr(MLIRContext *context, ArrayRef<Value> threadGrid,
+                           PatternBenefit benefit = 1)
+      : OpDistributionPattern<OpTy>(context, benefit), threadGrid(threadGrid) {}
+
+  VectorValue accessMemory(OpTy xferOp, VectorValue accumulator,
+                           LayoutAttr vectorLayout,
+                           PatternRewriter &rewriter) const {
+    // We need to take special consideration of the permutation map when
+    // lowering. When accessing memory, we use the memoryLayout, because that
+    // is how the data is accessed in memory. The data is stored in the vector
```

Comment:
To clarify, the memoryLayout is essentially the vectorLayout with the permutation map of the transfer_read/write applied to the vectorLayout?

Reply:
If so, does that mean on GPUs that have read/write instructions with permutations, we would use the vectorLayout instead of the memoryLayout?

Reply:
Yes, we would use vectorLayout there. This is just for the general case.

Reply:
Thanks.